status-im/status-react
# feat: add support for updating key-pairs and accounts from app signal

[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

fixes #20416

### Summary

* This PR attempts to update the signal processing of the mobile app to handle processing key-pair related signals for updating the key-pairs and accounts data in the app-db.
  * Functionality for adding new key-pairs / accounts via sync, editing accounts, and removing key-pairs / accounts via sync

### Testing notes

In order to test these changes the testing environment will need to have:
* An account with two devices that are synced to that account
* Multiple key pairs and accounts created and imported on both devices
* The key pairs and accounts feature flag enabled, until #20464 is merged

#### Platforms


- Android
- iOS

#### Areas that maybe impacted


##### Functional

- Editing accounts
- Removing accounts
- Missing key pairs and accounts sync

### Steps to test


Reproducing the intended behaviour can be seen in the screen captures provided in this PR.
Though here are some brief steps to reproduce each case:
* Editing the default status account
  * While viewing the Status mobile app on Device-A, use Device-B to edit the emoji or name of the default status account and save.
  * This test case is special because we need to handle default key-pair differently than other key pairs because of the associated `chat` account.
* Renaming account
  * While viewing the Status mobile app on Device-A, use Device-B to edit the emoji or name of an account and save.
* Adding new account to existing key pair
  * While viewing the Status mobile app on Device-A, use Device-B to add a new account to an existing key pair.
* Removing account from pair with multiple accounts
  * While viewing the Status mobile app on Device-A, use Device-B to remove an account from an existing key pair with multiple accounts.
* Adding new account and missing key pair
  * While viewing the Status mobile app on Device-A, use Device-B to create a new account and a new key pair. If the devices are synced, we should see the new account and key pair appear as a "missing" key pair.
* Removing account and missing key pair
  * While viewing the Status mobile app on Device-A, use Device-B to remove an account that is associated to a missing key pair. If the devices are synced, we should see the missing account and key pair disappear. 
* Removing account and existing key pair
  * While viewing the Status mobile app on Device-A, use Device-B to remove an account that is associated to an existing key pair. If the devices are synced, we should see the account and key pair disappear. 

### Before and after screenshots comparison

| Title | Screen Capture |
|---|---|
| Renaming account | <video src="https://github.com/status-im/status-mobile/assets/2845768/6311e821-52f3-4f8a-98ea-7f76345cc677" /> |
| Editing the default status account | <video src="https://github.com/status-im/status-mobile/assets/2845768/431d9880-c2eb-4e05-a6cb-a891330ab38b" /> |
| Adding new account to existing key pair | <video src="https://github.com/status-im/status-mobile/assets/2845768/51e6aaf1-62c0-449b-be39-6afb0ebcb359" /> |
| Removing account from key pair with multiple accounts | <video src="https://github.com/status-im/status-mobile/assets/2845768/33262928-a265-4b61-a34b-9221b120b40f" /> |
| Adding new account and missing key pair | <video src="https://github.com/status-im/status-mobile/assets/2845768/4fd57fad-0d2e-4fb0-ad49-9a81f914cc7b" /> |
| Removing account and missing key pair | <video src="https://github.com/status-im/status-mobile/assets/2845768/3bad4c49-aafa-4edd-871f-1dd8375e9765" /> |
| Removing account and existing key pair | <video src="https://github.com/status-im/status-mobile/assets/2845768/f23c7950-bd11-46bf-9826-922be30328b7" /> |

status: ready 

### Risk

Described potential risks and worst case scenarios.

Tick **one**:
